stdlib/Artifacts/src/Artifacts.jl

```diff
@@ -605,14 +607,17 @@ macro artifact_str(name, platform=nothing)
     # Invalidate calling .ji file if Artifacts.toml file changes
     Base.include_dependency(artifacts_toml)
 
+    # Check if the user has run `using Pkg.Artifacts: ensure_artifact_installed`, and thus supports lazy artifacts
+    ensure_artifact_installed = isdefined(__module__, :ensure_artifact_installed) ? GlobalRef(__module__, :ensure_artifact_installed) : nothing
```

Comment:
Instead of looking for `:ensure_artifact_installed` I think it's better to look for `:Pkg`, so that users can use `import Pkg` to avoid polluting their namespace.

Reply:
I did that initially, but eventually felt this polluted the namespace the least and gives the most forward-compatibility. `using Pkg.Artifacts` is a non-starter (since it'll conflict with `using Artifacts`), while `Pkg` is just a rather short name.

Reply:
`ensure_artifact_installed` is a private API, and we can change that name at any time.  It would be better to ask the user to import something that will not change, like `Pkg`.

Reply:
Wait, I'm wrong, `ensure_artifact_installed` is not private, it's exported.  I still think it's better (in the event that we switch names in the future, after an appropriate deprecation period) to depend on `Pkg` though.

Reply:
> Wait, I'm wrong, ensure_artifact_installed is not private, it's exported. 

Well, it should really have been private.

Reply:
Agreed.  It's not exported from `Pkg`, but it is exported from `Artifacts` because it's used in other parts of `Pkg`.

Reply:
Writing `using Pkg: Pkg` also pollutes the namespace, but uses a more common variable name than `using Pkg: Artifacts.ensure_artifact_installed` uses. Since the existing `lazy` feature here already means that changing the name of this would be breaking (once we release Artifacts v1.6 and users are depending on it), so you've already committed yourself to having `Pkg` provide this function. So the question here is whether you want the API to be requiring the existence of a global variable named `Pkg` to signal the existence of `Pkg.Artifacts`and `Pkg.Artifacts.ensure_artifact_installed` (with possibility for there to be additional functions added later that this macro may select instead), or whether the user should be more explicit about the name `ensure_artifact_installed` (with possibility that the user can provide other implementations of this function in the future, and thus aren't forced to always depend upon Pkg in the future).

Or as a combination, we could rename `Pkg.Artifacts` to `Pkg.ArtifactsDownloader` (or some other name of your liking), so that it doesn't conflict with the name of this package, and thus gives you the flexibility to make this also a toplevel package in the future (of the same name, or with `using ArtifactsNext: ArtifactsNext as ArtifactsDownloader`)

Reply:
To me it's a clear choice: looking at `Pkg` gives something that users will most likely already have (much more likely than `ensure_artifact_installed`, for instance) and also gives us the flexibility in code to polymorph to future `Pkg` releases where we re-organize things.  It pushes the work of backwards-compatibility onto `Artifacts` instead of onto user code, which is always a good thing when it has no performance impact.